include/fiction/io/write_operational_domain.hpp

```diff
@@ -145,32 +145,33 @@ inline void write_operational_domain(const OpDomain& opdom, std::ostream& os,
                               "critical temperature");
         }
 
-        for (const auto& [sim_param, op_val] : opdom.domain_values)
-        {
-            // skip non-operational samples if the respective flag is set
-            if (params.writing_mode == write_operational_domain_params::sample_writing_mode::OPERATIONAL_ONLY &&
-                std::get<0>(op_val) == operational_status::NON_OPERATIONAL)
+        opdom.for_each(
+            [&params, &writer, &num_dimensions](const auto& sim_param, const auto& op_val)
             {
-                continue;
-            }
+                // skip non-operational samples if the respective flag is set
+                if (params.writing_mode == write_operational_domain_params::sample_writing_mode::OPERATIONAL_ONLY &&
+                    std::get<0>(op_val) == operational_status::NON_OPERATIONAL)
+                {
+                    return;
+                }
 
-            const auto tag = std::get<0>(op_val) == operational_status::OPERATIONAL ? params.operational_tag :
-                                                                                      params.non_operational_tag;
+                const auto tag = std::get<0>(op_val) == operational_status::OPERATIONAL ? params.operational_tag :
+                                                                                          params.non_operational_tag;
 
-            if (num_dimensions == 1)
-            {
-                writer.write_line(sim_param.parameters[0], tag, std::get<1>(op_val));
-            }
-            else if (num_dimensions == 2)
-            {
-                writer.write_line(sim_param.parameters[0], sim_param.parameters[1], tag, std::get<1>(op_val));
-            }
-            else  // num_dimensions == 3
-            {
-                writer.write_line(sim_param.parameters[0], sim_param.parameters[1], sim_param.parameters[2], tag,
-                                  std::get<1>(op_val));
-            }
-        }
+                if (num_dimensions == 1)
+                {
+                    writer.write_line(sim_param.parameters[0], tag, std::get<1>(op_val));
+                }
+                else if (num_dimensions == 2)
+                {
+                    writer.write_line(sim_param.parameters[0], sim_param.parameters[1], tag, std::get<1>(op_val));
+                }
+                else  // num_dimensions == 3
+                {
+                    writer.write_line(sim_param.parameters[0], sim_param.parameters[1], sim_param.parameters[2], tag,
+                                      std::get<1>(op_val));
+                }
```

Comment:
I know that this code isn't new to this PR, but I just wonder whether we should spell out the first three dimensions here explicitly with an `else` that throws an error. Would make it more robust against future changes and would remind us to fix it in time rather than providing wrong output.

Analogous for the other domain writers.

Reply:
good idea!